beacon-chain/db/filesystem/blob.go

```diff
@@ -76,138 +80,152 @@ func NewBlobStorage(opts ...BlobStorageOption) (*BlobStorage, error) {
-func (bs *BlobStorage) Save(sidecar blocks.VerifiedROBlob) error {
-	startTime := time.Now()
-	fname := namerForSidecar(sidecar)
-	sszPath := fname.path()
-	exists, err := afero.Exists(bs.fs, sszPath)
-	if err != nil {
-		return err
-	}
-	if exists {
-		log.WithFields(logging.BlobFields(sidecar.ROBlob)).Debug("Ignoring a duplicate blob sidecar save attempt")
-		return nil
-	}
-	if bs.pruner != nil {
-		if err := bs.pruner.notify(sidecar.BlockRoot(), sidecar.Slot(), sidecar.Index); err != nil {
-			return errors.Wrapf(err, "problem maintaining pruning cache/metrics for sidecar with root=%#x", sidecar.BlockRoot())
+// If any blob storage directories are found for layouts besides the configured layout, migrate them.
+func (bs *BlobStorage) migrateLayouts() error {
+	for _, name := range LayoutNames {
+		if name == bs.layoutName {
+			continue
+		}
+		from, err := newLayout(name, bs.fs, bs.cache, nil)
+		if err != nil {
+			return err
+		}
+		if err := migrateLayout(bs.fs, from, bs.layout, bs.cache); err != nil {
+			return errors.Wrapf(err, "failed to migrate layout from %s to %s", name, bs.layoutName)
 		}
 	}
+	return nil
+}
 
+func (bs *BlobStorage) writePart(sidecar blocks.VerifiedROBlob) (string, error) {
+	ident := identForSidecar(sidecar)
 	// Serialize the ethpb.BlobSidecar to binary data using SSZ.
 	sidecarData, err := sidecar.MarshalSSZ()
 	if err != nil {
-		return errors.Wrap(err, "failed to serialize sidecar data")
+		return "", errors.Wrap(err, "failed to serialize sidecar data")
 	} else if len(sidecarData) == 0 {
-		return errSidecarEmptySSZData
+		return "", errSidecarEmptySSZData
 	}
 
-	if err := bs.fs.MkdirAll(fname.dir(), directoryPermissions); err != nil {
-		return err
+	if err := bs.fs.MkdirAll(bs.layout.dir(ident), directoryPermissions); err != nil {
+		return "", err
 	}
-	partPath := fname.partPath(fmt.Sprintf("%p", sidecarData))
-
-	partialMoved := false
-	// Ensure the partial file is deleted.
-	defer func() {
-		if partialMoved {
-			return
-		}
-		// It's expected to error if the save is successful.
-		err = bs.fs.Remove(partPath)
-		if err == nil {
-			log.WithFields(logrus.Fields{
-				"partPath": partPath,
-			}).Debugf("Removed partial file")
-		}
-	}()
+	partPath := bs.layout.partPath(ident, fmt.Sprintf("%p", sidecarData))
 
 	// Create a partial file and write the serialized data to it.
 	partialFile, err := bs.fs.Create(partPath)
 	if err != nil {
-		return errors.Wrap(err, "failed to create partial file")
+		return "", errors.Wrap(err, "failed to create partial file")
 	}
 
 	n, err := partialFile.Write(sidecarData)
 	if err != nil {
 		closeErr := partialFile.Close()
 		if closeErr != nil {
-			return closeErr
+			return partPath, closeErr
 		}
-		return errors.Wrap(err, "failed to write to partial file")
+		return partPath, errors.Wrap(err, "failed to write to partial file")
 	}
 	if bs.fsync {
 		if err := partialFile.Sync(); err != nil {
-			return err
+			return partPath, err
```

Comment:
do we need to close partial File here?

Reply:
Yes - apparently a bug we had before (this func is basically a copy/paste to reduce cyclomatic complexity complaint on `Save`). Will deal with Close in a defer.

Reply:
@prestonvanloon since I'm cleaning this up - I noticed that we check:
1. that the sidecar data len is not equal to zero
2. that the number of bytes written is equal to the sidecar data len
3. that the number of bytes written is not equal to zero

But 3 is impossible if 1 and 2 checks don't fail. So I'm going to delete 3.

Reply:
updated